Source/WebGPU/WGSL/Overload.cpp

```diff
@@ -363,10 +362,83 @@ bool OverloadResolver::unify(const AbstractValue& parameter, unsigned argumentVa
     return *resolvedValue == argumentValue;
 }
 
-void OverloadResolver::assign(TypeVariable variable, Type* type, Substitution::Inferred inferred)
+bool OverloadResolver::assign(TypeVariable variable, Type* type, Substitution::Inferred inferred)
 {
     logLn("assign ", variable, " => ", *type);
+    if (variable.constraints) {
+        auto* primitive = std::get_if<Types::Primitive>(type);
+        if (!primitive)
+            return false;
+
+        switch (primitive->kind) {
+        case Types::Primitive::AbstractInt:
+            if (variable.constraints < TypeVariable::AbstractInt)
+                return false;
+            if (variable.constraints & TypeVariable::AbstractInt)
+                break;
+            if (variable.constraints & TypeVariable::I32) {
+                type = m_types.i32Type();
+                break;
+            }
+            if (variable.constraints & TypeVariable::U32) {
+                type = m_types.u32Type();
+                break;
+            }
+            if (variable.constraints & TypeVariable::AbstractFloat) {
+                type = m_types.abstractFloatType();
+                break;
+            }
+            if (variable.constraints & TypeVariable::F32) {
+                type = m_types.f32Type();
+                break;
+            }
+            if (variable.constraints & TypeVariable::F16) {
+                // FIXME: Add F16 support
+                // https://bugs.webkit.org/show_bug.cgi?id=254668
+                RELEASE_ASSERT_NOT_REACHED();
+            }
+            RELEASE_ASSERT_NOT_REACHED();
+        case Types::Primitive::I32:
+            if (!(variable.constraints & TypeVariable::I32))
+                return false;
+            break;
+        case Types::Primitive::U32:
+            if (!(variable.constraints & TypeVariable::U32))
+                return false;
+            break;
+        case Types::Primitive::AbstractFloat:
+            if (variable.constraints < TypeVariable::AbstractFloat)
+                return false;
+            if (variable.constraints & TypeVariable::AbstractFloat)
+                break;
+            if (variable.constraints & TypeVariable::F32) {
+                type = m_types.f32Type();
+                break;
+            }
+            if (variable.constraints & TypeVariable::F16) {
+                // FIXME: Add F16 support
+                // https://bugs.webkit.org/show_bug.cgi?id=254668
+                RELEASE_ASSERT_NOT_REACHED();
+            }
+            RELEASE_ASSERT_NOT_REACHED();
+        case Types::Primitive::F32:
+            if (!(variable.constraints & TypeVariable::F32))
+                return false;
+            break;
+        // FIXME: Add F16 support
+        // https://bugs.webkit.org/show_bug.cgi?id=254668
```

Comment:
Is fp16 special somehow? I don't think it is. It's supported on all Metal devices as a first-class citizen. Seems like there's no reason not to just do it now 🤔